servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/AutomaticRetryFilter.java

```diff
@@ -31,55 +27,31 @@
 import io.servicetalk.http.api.StreamingHttpRequest;
 import io.servicetalk.http.api.StreamingHttpRequester;
 import io.servicetalk.http.api.StreamingHttpResponse;
-import io.servicetalk.transport.api.RetryableException;
-
-import static io.servicetalk.concurrent.api.Completable.failed;
-import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
 
 /**
  * A {@link StreamingHttpClient} filter that will account for transient failures introduced by a {@link LoadBalancer}
  * not being ready for {@link #request(HttpExecutionStrategy, StreamingHttpRequest)} and retry/delay requests until the
  * {@link LoadBalancer} is ready.
  */
-final class LoadBalancerReadyStreamingHttpClientFilter extends StreamingHttpClientFilter {
-    private final LoadBalancerReadySubscriber loadBalancerReadySubscriber;
-    private final int maxRetryCount;
+final class AutomaticRetryFilter extends StreamingHttpClientFilter {
```

Comment:
It looks very similar to `RetryingHttpRequesterFilter` now. And `RetryingHttpRequesterFilter` already handles `RetryableConnectException` by default.
It's harder for users to understand the difference between using `DefaultAutomaticRetryStrategyProvider.Builder` and `RetryingHttpRequesterFilter.Builder`. And it's harder to discover that `RetryingHttpRequesterFilter` exists when `AutomaticRetryStrategyProvider` exposed on the builder.

I see that `AutomaticRetryStrategyProvider` needs access to `LoadBalancer`. Could we explore an opportunity to merge it with `RetryingHttpRequesterFilter`?

Reply:
I agree that they are similar in terms of the fact that they are doing retries, however, the type of retries they do is very different:

- _auto-retry_: This is for expected errors in regular paths when due to races in closing and selecting a connection, we select a bad connection. Such errors are always retryable and since they do not write anything on the network, there is no need to add backoff, jitter, etc.
- _user-retries_: This is for failures which happen during client processing or from the server. Since there is work done on the network, one needs to introduce jitters, backoff, etc.

From a functionality wise, auto-retry also needs the `LoadBalancer` as you noticed. So, I believe they are different enough to be configured differently.

Reply:
Agreed, and as discussed offline, it's also better to have auto-retries closer to transport, when user-retries may handle more logic from other filters.